dataframe/time_series.go

```diff
@@ -0,0 +1,369 @@
+			uniqueLabelKeys[k] = struct{}{}
+		}
+	}
+
+	// Sort things for more deterministic output
+	sort.Strings(uniqueValueNames)
+	sortedUniqueLabelKeys := []string{}
+	for k := range labelKeyToWideIndices {
+		sortedUniqueLabelKeys = append(sortedUniqueLabelKeys, k)
+	}
+	sort.Strings(sortedUniqueLabelKeys)
+	uniqueFactorNames := make([]string, 0, len(uniqueLabelKeys))
+	for k := range uniqueLabelKeys {
+		uniqueFactorNames = append(uniqueFactorNames, k)
+	}
+	sort.Strings(uniqueFactorNames)
+
+	// build new Frame with new schema
+	longFrame := New(wideFrame.Name, // time , value fields (numbers)..., factor fields (strings)...
+		NewField(wideFrame.Fields[tsSchema.TimeIndex].Name, nil, []time.Time{})) // time field is first field
+
+	i := 1
+	valueNameToLongFieldIdx := map[string]int{} // valueName -> field index of longFrame
+	for _, name := range uniqueValueNames {
+		longFrame.Fields = append(longFrame.Fields, &Field{ // create value (number) vectors
+			Name:   name,
+			Vector: NewVectorFromPType(uniqueValueNamesToType[name], 0),
+		})
+		valueNameToLongFieldIdx[name] = i
+		i++
+	}
+
+	factorNameToLongFieldIdx := map[string]int{} // label Key -> field index for label value of longFrame
+	for _, name := range uniqueFactorNames {
+		longFrame.Fields = append(longFrame.Fields, NewField(name, nil, []string{})) // create factor fields
+		factorNameToLongFieldIdx[name] = i
+		i++
+	}
+
+	// Populate data of longFrame from wideframe
+	longFrameCounter := 0
+	for wideRowIdx := 0; wideRowIdx < wideLen; wideRowIdx++ { // loop over each row of wideFrame
+		time, ok := wideFrame.ConcreteAt(tsSchema.TimeIndex, wideRowIdx)
+		if !ok {
+			return nil, fmt.Errorf("time may not have nil values")
+		}
+		for _, labelKey := range sortedUniqueLabelKeys {
+			longFrame.Extend(1) // grow each Fields's vector by 1
+			longFrame.Set(0, longFrameCounter, time)
+
+			for i, wideFieldIdx := range labelKeyToWideIndices[labelKey] {
+				wideField := wideFrame.Fields[wideFieldIdx]
+				if i == 0 {
+					for k, v := range wideField.Labels {
+						longFrame.Set(factorNameToLongFieldIdx[k], longFrameCounter, v)
+					}
+				}
+				longValueFieldIdx := valueNameToLongFieldIdx[wideField.Name]
+				longFrame.Set(longValueFieldIdx, longFrameCounter, wideFrame.CopyAt(wideFieldIdx, wideRowIdx))
+
+			}
+
+			longFrameCounter++
+		}
+	}
+
+	return longFrame, nil
+}
+
+// TimeSeriesSchema is information about a Dataframe's schema.  It is populated from
+// the Frame's TimeSeriesSchema() method.
+type TimeSeriesSchema struct {
+	Type           TimeSeriesType // the type of series, as determinted by frame.TimeSeriesSchema()
+	TimeIndex      int            // Field index of the time series index
+	TimeIsNullable bool           // true if the time index is nullable (of *time.Time)
+	ValueIndices   []int          // Field indices of Number type columns, as determinted by NumericVectorPTypes
+	FactorIndices  []int          // Field indices of string or *string Fields
```

Comment:
What happens with "other" fields?  do they have to be either numbers or strings?

Reply:
Changed, other fields (not string, and not the TimeIndex) are now all ValueIndicies.